node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,164 @@
+    /// or a `ValueOrChunk::ChunkWithProof` if it is. In the latter case it will return only the
+    /// `chunk_index`-th chunk of the value's byte representation.
+    pub fn new(data: V, chunk_index: u64) -> Result<Self, ChunkingError>
+    where
+        V: Serialize,
+    {
+        let bytes = bincode::serialize(&data).map_err(ChunkingError::SerializationError)?;
+        // NOTE: Cannot accept the chunk size bytes as an argument without changing the
+        // IndexedMerkleProof. The chunk size there is hardcoded and will be used when
+        // determining the chunk.
+        if bytes.len() <= ChunkWithProof::CHUNK_SIZE_BYTES {
+            Ok(ValueOrChunk::Value(data))
+        } else {
+            let chunk_with_proof = ChunkWithProof::new(&bytes, chunk_index)
+                .map_err(ChunkingError::MerkleConstruction)?;
+            Ok(ValueOrChunk::ChunkWithProof(chunk_with_proof))
+        }
+    }
+}
+
+/// Represents an enum that can contain either a whole trie or a chunk of it.
+pub type TrieOrChunk = ValueOrChunk<TrieRaw>;
+
+impl<V: Debug> Display for ValueOrChunk<V> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            ValueOrChunk::Value(data) => f.debug_tuple("Value").field(data).finish(),
+            ValueOrChunk::ChunkWithProof(chunk) => f
+                .debug_struct("ChunkWithProof")
+                .field("index", &chunk.proof().index())
+                .field("hash", &chunk.proof().root_hash())
+                .finish(),
+        }
+    }
+}
+
+/// Represents the ID of a `TrieOrChunk` - containing the index and the root hash.
+/// The root hash is the hash of the trie node as a whole.
+/// The index is the index of a chunk if the node's size is too large and requires chunking. For
+/// small nodes, it's always 0.
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct TrieOrChunkId(pub u64, pub Digest);
+
+impl TrieOrChunkId {
+    /// Returns the trie key part of the ID.
+    pub fn digest(&self) -> &Digest {
+        &self.1
+    }
+
+    /// Given a serialized ID, deserializes it for display purposes.
+    fn fmt_serialized(f: &mut Formatter, serialized_id: &[u8]) -> fmt::Result {
+        match bincode::deserialize::<Self>(serialized_id) {
+            Ok(ref trie_or_chunk_id) => Display::fmt(trie_or_chunk_id, f),
+            Err(_) => f.write_str("<invalid>"),
+        }
+    }
+}
+
+/// Helper struct to on-demand deserialize a trie or chunk ID for display purposes.
+pub(crate) struct TrieOrChunkIdDisplay<'a>(pub &'a [u8]);
+
+impl<'a> Display for TrieOrChunkIdDisplay<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        TrieOrChunkId::fmt_serialized(f, self.0)
+    }
+}
+
+impl Display for TrieOrChunkId {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "({}, {})", self.0, self.1)
+    }
+}
+
+#[cfg(test)]
+mod value_chunking {
```

Comment:
I believe this should say `mod tests`.
We could consider introducing a submodule called `value_chunking` to group the tests, but it's something we avoided so far in the `casper-node`. Also, this is just a single test so there's no need for sophisticated grouping :)

Reply:
Why should it say `mod tests`? Is it a convention?

Reply:
On the contrary to the requirement for the integration tests folder to be named `tests`, this one is not a strict requirement. It's more like de-facto standard.

Reply:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)